### PR TITLE
Add cache support to decoding journey

### PR DESCRIPTION
**Pull request by chenmoneygithub**
There are a few things to note here:
- XLA path is not supported as it returns a different result from non-xla graph, I am looking into the cause.
- Beam search does not support cache for now, because it needs to maintain `num=num_beams` cache. I will do this as a followup.
- We are using a lot of slicing and update calls now, unsure how to optimize it. I have been looking into this for too long, so I am trapped in the rabbit hole.
- To play with the API and compare the performance, you can use this [colab](https://colab.research.google.com/gist/chenmoneygithub/b02eaf70d3004063b1c4b92965174ec6/cached-decoding-solution-branch-version.ipynb).

This is oddly and insanely complex, so thanks for the review work. 

Cheers.

### CODE REVIEW

**Comment by mattdangerw** [comment]
Thanks for the PR! This is some of the most complex code we have added yet! Went deep into this and played around with the code for most of yesterday.

Short version, I have a commit based on your branch we may want to just fold into the PR -> https://github.com/keras-team/keras-nlp/commit/11181a1442d8532013f3d0a55b8e1f48be8efd64. It should be both a little more efficient, and cuts down on code a lot.

Some bigger notes:

- Let's move all of the new forward pass to `GptCausalLM`. There is really no use case for this cache stuff besides auto-regressive decoding, so we can limit the "blast radius" of the change significantly by leaving `call_with_cache` only on the task model and not the backbone.
- Rather then slice the input short after doing token and position embeddings, we should actually keep the input short coming in to the model. Basically `"token_ids"` during generation should have shape `(bs, 1)`. There are a few reasons for this. First, it is more efficient, in the version you have here you are recomputing the same token and position embeddings over and over. Second, it allows for much cleaner state passing in the sampling code. This requires some code changes for position embedding and mask computation.
- Let's only support a "cached" decoding path right now. Support both cached and non-cached makes the code really hard to follow, and I don't think there is a good case for it.
- Let's avoid passing unnecessary state between the sampler and masked lm classes, e.g. `initial_output` and `existing_outputs`.

Some more minor notes:

- Let's avoid a dynamic return types where we don't have to. I think `TransformerDecoder` has to support a return with a cache or not (for compatibility). But the cached attention layer can just always return the cache (None or not).
- Let's remove what options we can from `CachedMultiHeadAttention` and keep it simple, we don't need 1-1 parity with `MultiHeadAttention` (that class is complex).
- We are using `dynamic_update_slice` to do simple stacking and unstacking of the cache tensor. Let's just used `tf.stack` and `tf.unstack`. Much more readable.
- Similarly in the sampler we use `scatten_nd` op in the sampler to do the same thing as `dynamic_update_slice`. Let's just use `dynamic_update_slice` for readability.
- I thought the names `cache` and `cache_index` were nicely parallel as inputs, the reader will know they are linked!

Open questions:

 - Let's try to keep figuring out XLA compilation failures here. We know it is doable!
 - With the cache only approach, beam search in it's current form is pretty useless. How much work is adding support for beam search?

**Comment by chenmoneygithub** [comment]
@mattdangerw Thanks Matt!

> Let's move all of the new forward pass to GptCausalLM. 
 Putting things into `GptCausalLM` sounds good to me. 

> Rather then slice the input short after doing token and position embeddings...
I think I am already doing that, the only exception is the first pass to build the cache. I checked your commit, it looks good to me to combine the first pass as well.

> Let's only support a "cached" decoding path right now. 
I believe the "cached" path is broken on TPU due to XLA issue, so I guess we should keep the non-cached one until we solve the XLA issue. BTW, "cache" is not guaranteed to be faster, and actually slower when the `max_length` is short.

> Let's avoid passing unnecessary state between the sampler and masked lm classes, e.g. initial_output and existing_outputs.
They are actually required for some decoding algorithm, such as contrastive search. Basically these decoding algo needs to know previous token's representation.

> Let's avoid a dynamic return types where we don't have to.
Fine with me!

> Let's remove what options we can from CachedMultiHeadAttention and keep it simple..
I am not sure if we want to diverge from `MultiHeadAttention`, because techinically this layer is a addon to MHA. 

> We are using dynamic_update_slice to do simple stacking and unstacking of the cache tensor.
I think we need to keep cache of a static shape? `dynamic_update_slice` is actually not that bad IMO, it's much more straightforward than `tf.tensor_scatter_nd_update`.

> Similarly in the sampler we use scatten_nd op in the sampler to do the same thing as dynamic_update_slice
SGTM!

> I thought the names cache and cache_index were nicely parallel as inputs, the reader will know they are linked!
good point, SGTM!

> Let's try to keep figuring out XLA compilation failures here. We know it is doable!
Sure, but I don't want this to be blocking. HF proves it's doable in their way, but we may not want to mirror their coding logic.

> With the cache only approach, beam search in it's current form is pretty useless. How much work is adding support for beam search?
Holy a lot T T....

**Comment by mattdangerw** [comment]
> BTW, "cache" is not guaranteed to be faster, and actually slower when the max_length is short.

How short practically? Is it enough that real world generative use cases will want the non-cached version in practice? If we can scope down (at least temporarily) here, that would make this much easier to land.

>> We are using dynamic_update_slice to do simple stacking and unstacking of the cache tensor.

> I think we need to keep cache of a static shape? dynamic_update_slice is actually not that bad IMO, it's much more straightforward than tf.tensor_scatter_nd_update.

Static shape is actually the only thing a unstack/stack combo will handle. Take a look at the commit I linked, it's a pure readability and code reduction. Logic untouched.

>> Let's try to keep figuring out XLA compilation failures here. We know it is doable!

> Sure, but I don't want this to be blocking. HF proves it's doable in their way, but we may not want to mirror their coding logic.

The issue is our whole setup is for XLA right now (all of our padding to static shapes). If we don't support XLA, we actually have a overly complex and underly efficient version right now compared to what is "non XLA compilable." Take a look at the original GPT implementation for a non-XLA version that uses `concat` and dynamic shapes that is both simpler and faster than ours.

Let's keep digging into XLA failures!

**Comment by mattdangerw** [comment]
> > Let's avoid passing unnecessary state between the sampler and masked lm classes, e.g. initial_output and existing_outputs.

> They are actually required for some decoding algorithm, such as contrastive search. Basically these decoding algo needs to know previous token's representation.

This is an interesting point we should discuss. If the decoding algorithm needs it, seems like that could still not leak into `GPT2MaskedLM` implementation. If would be very doable to write a sampler that used both a previous token and next token logits, without that logic needing to spread across to our language model classes as well.

File: keras_nlp/models/gpt2/gpt2_backbone.py [hunk 1]
```diff
@@ -170,6 +174,116 @@ def __init__(
...
+        else:
+            position_embedding = self.position_embeddings.position_embeddings[
+                current_index, :
+            ]
+        x = token_embedding + position_embedding
+        x = self.embeddings_dropout(x)
+        if current_index is not None:
+            batch_size = tf.shape(x)[0]
+            hidden_dim = tf.shape(x)[2]
+            x = tf.slice(x, [0, current_index, 0], [batch_size, 1, hidden_dim])
```
Comment (chenmoneygithub) [inline keras_nlp/models/gpt2/gpt2_backbone.py:220]: @mattdangerw This line makes it XLA-compatible

File: keras_nlp/layers/cached_multi_head_attention.py [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+        cache: the updated cache.
+    """
+
+    def call(
+        self,
+        query,
+        value,
+        key=None,
+        attention_mask=None,
+        use_causal_mask=True,
```
Comment (mattdangerw) [inline keras_nlp/layers/cached_multi_head_attention.py:66]: This feels weird to me, adding a cache should not change the default masking behavior.

Could we just remove this option and fixup the unit tests?

File: keras_nlp/layers/cached_multi_head_attention_test.py [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+        layer(query=x, value=x)
+
+    def test_cache_call_is_correct(self):
+        batch_size = 2
+        seq_len = 5
+        num_heads = 2
+        key_dim = 4
+        layer = CachedMultiHeadAttention(num_heads=num_heads, key_dim=key_dim)
+        x = tf.random.uniform(shape=[batch_size, seq_len, num_heads * key_dim])
+        # [2, batch_size, seq_length, num_heads, key_dim]
```
Comment (mattdangerw) [inline keras_nlp/layers/cached_multi_head_attention_test.py:38]: this comment is just the same as the line below

File: keras_nlp/layers/transformer_decoder.py [hunk 4]
```diff
@@ -208,6 +211,8 @@ def call(
         decoder_attention_mask=None,
         encoder_padding_mask=None,
         encoder_attention_mask=None,
+        cache=None,
```
Comment (mattdangerw) [inline keras_nlp/layers/transformer_decoder.py:214]: we need to document these and the extra output, especially as TransformerEncoder/TransformerDecoder are two of the main offerings from our layers.

File: keras_nlp/models/gpt2/gpt2_backbone.py [hunk 5]
```diff
@@ -144,7 +145,10 @@ def __init__(
                 kernel_initializer=_gpt_2_kernel_initializer(stddev=0.02),
                 normalize_first=True,
                 name=f"transformer_layer_{i}",
-            )(x, decoder_padding_mask=padding_mask)
```
Comment (mattdangerw) [inline keras_nlp/models/gpt2/gpt2_backbone.py:147]: is there a reason for this change? actually found it more readable as before

File: keras_nlp/models/gpt2/gpt2_backbone.py [hunk 6]
```diff
@@ -170,6 +174,116 @@ def __init__(
         self.intermediate_dim = intermediate_dim
         self.dropout = dropout
         self.max_sequence_length = max_sequence_length
+        self.transformer_layers = []
```
Comment (mattdangerw) [inline keras_nlp/models/gpt2/gpt2_backbone.py:177]: I think rather than add a bunch of attributes, we should just move the `self.get_layer("name")` calls into the call_with_cache method.

Otherwise we are playing a dangerous game, as adding a layer that is not a toplevel of the functional model to self will break all our checkpoints!

File: keras_nlp/models/gpt2/gpt2_backbone.py [hunk 7]
```diff
@@ -170,6 +174,116 @@ def __init__(
...
+        for i in range(self.num_layers):
+            self.transformer_layers.append(
+                self.get_layer(f"transformer_layer_{i}")
+            )
+        self.layer_norm = self.get_layer("layer_norm")
+        self.token_embeddings = self.get_layer("token_embedding")
+        self.position_embeddings = self.get_layer("position_embedding")
+        self.embeddings_dropout = self.get_layer("embeddings_dropout")
+
+    def call_with_cache(self, inputs, cache, cache_index=None):
```
Comment (mattdangerw) [inline keras_nlp/models/gpt2/gpt2_backbone.py:187]: I think it might be simpler to just blow up all the inputs here as explicit args, e.g `token_ids, padding_mask, cache, cache_index`

Mixing a dict for some inputs and not for others is a bit confusing.

File: keras_nlp/models/gpt2/gpt2_causal_lm.py [hunk 8]
```diff
@@ -200,17 +201,57 @@ def backbone_cls(cls):
...
+            self.backbone.token_embedding.embeddings,
+            transpose_b=True,
+        )
+        return output, cache
+
+    def _get_token_probability(
+        self,
+        prompt,
+        mask,
+        cache_index=None,
```
Comment (mattdangerw) [inline keras_nlp/models/gpt2/gpt2_causal_lm.py:228]: nit: flip the order of these args, so `cache` then `cache_index`

File: keras_nlp/samplers/beam_sampler.py [hunk 9]
```diff
@@ -85,13 +86,26 @@ def get_next_token(self, next_token_probs):
...
+        from_logits=True,
+        cache=None,
+        token_probs=None,
     ):
         """Sampling logic implementation.
 
         Because beam search uses a different loop body, we have to override the
         whole `sample` method instead of just the `get_next_token` method.
         """
+        if cache is not None:
```
Comment (mattdangerw) [inline keras_nlp/samplers/beam_sampler.py:103]: How do we want to handle this? Should we just unexport beam for now, and leave cache stuff as a follow up?

File: keras_nlp/samplers/sampler.py [hunk 10]
```diff
@@ -39,6 +40,17 @@
     from_logits: bool, defaults to True. Indicate if the `token_probability_fn`
         returns logits. If False, `token_probability_fn` returns probability
         distributions.
+    cache: a dense int tensor, the cache used in decoding. The cache
+        stores the K and V of each
```
Comment (mattdangerw) [inline keras_nlp/samplers/sampler.py:44]: K, V -> key, value

Also , rather than point to `CachedMultiHeadAttention`, I would give a high level description of the function of the cache here.

```
A cache of intermediate key and value tensor computed by each
decoder self-attention layer. These values will only be computed
once for each new token in the generated sequence.
```

**Review by mattdangerw — CHANGES_REQUESTED** [review]
Just some small comments as we do the more major surgery talked about on the comment thread.

File: keras_nlp/layers/transformer_decoder.py (hunk 4 above)
Comment (mattdangerw) [reply keras_nlp/layers/transformer_decoder.py:214]: Still need to document return type!

File: keras_nlp/layers/cached_multi_head_attention.py [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+            else:
+                k_update_indices = [0, 0, cache_index, 0, 0]
+                v_update_indices = [1, 0, cache_index, 0, 0]
+            cache = dynamic_update_slice(
+                cache, key[tf.newaxis, ...], k_update_indices
+            )
+            cache = dynamic_update_slice(
+                cache, value[tf.newaxis, ...], v_update_indices
+            )
+            key = cache[0, :, : cache_index + 1, :, :]
```
Comment (mattdangerw) [inline keras_nlp/layers/cached_multi_head_attention.py:102]: Just a comment... It is still unclear to my why we need to slice here at all. Why not leave full length and rely on mask? I want to dig into this more.

File: keras_nlp/layers/cached_multi_head_attention.py [hunk 12]
```diff
@@ -0,0 +1,116 @@
...
+            attention_mask=attention_mask,
+            use_causal_mask=use_causal_mask,
+        )
+
+        query = self._query_dense(query)
+        key = self._key_dense(key)
+        value = self._value_dense(value)
+
+        if cache is not None:
+            if cache_index is None:
```
Comment (mattdangerw) [inline keras_nlp/layers/cached_multi_head_attention.py:88]: I think there are some changes we could make here that would make this less special cased to the way we are calling it.

 - Always pass both a `cache` and `cache_index` together. Either both are `None` or not.
 - On the assignment lines for `key` and `cache`, slice by `cache_index + seq_len`.

With these changes 0 will no longer be a special index. We always update the `cache` at `cache_index`, and the width of the update == the width of the input!

File: keras_nlp/layers/cached_multi_head_attention.py [hunk 13]
```diff
@@ -0,0 +1,116 @@
...
+                seq_len = tf.shape(query)[1]
+                k_update_indices = [0, 0, 0, 0, 0]
```
Comment (mattdangerw) [inline keras_nlp/layers/cached_multi_head_attention.py:90]: We could use the unstack stack trick here. Then you can have the same `update_indices` for each dynamic_update_slice.

File: keras_nlp/models/gpt2/gpt2_causal_lm.py [hunk 14]
```diff
@@ -188,6 +189,18 @@ def __init__(
         self.backbone = backbone
         self.preprocessor = preprocessor
 
+        # These layer attributes are used in cached decoding.
+        # We save them as class attrs to avoid excessive `get_layer` calls.
+        self.transformer_layers = []
```
Comment (mattdangerw) [inline keras_nlp/models/gpt2/gpt2_causal_lm.py:194]: We should remove these lines here, and call `self.backbone.get_layer("xx")` directly in the methods that need them.

The issue with adding these layers to self is you are actually fundamentally changing the structure of our checkpoints! Keras will consider these new layers though the `setattr` tracking on the base layer.

File: keras_nlp/models/gpt2/gpt2_causal_lm.py [hunk 15]
```diff
@@ -200,17 +213,138 @@ def backbone_cls(cls):
     def preprocessor_cls(cls):
         return GPT2CausalLMPreprocessor
 
-    def _get_token_probability(self, prompt, mask):
+    def call_with_cache(self, inputs, cache, cache_index=None):
```
Comment (mattdangerw) [inline keras_nlp/models/gpt2/gpt2_causal_lm.py:216]: Let's just move all the inputs into an input dict. So `inputs` a dict of `token_ids`, `padding_mask`, `cache`, `cache_index`.

Having two different levels where arguments are supplied is confusing. We want the mental model to be this is just a separate forward pass with new normal inputs added.

File: keras_nlp/models/gpt2/gpt2_causal_lm.py [hunk 16]
```diff
@@ -200,17 +213,138 @@ def backbone_cls(cls):
...
+        token_ids = inputs["token_ids"]
+        padding_mask = inputs["padding_mask"]
+        token_embedding = self.token_embeddings(token_ids)
+        if cache_index is None:
+            position_embedding = self.position_embeddings(token_embedding)
+        else:
+            position_embedding = self.position_embeddings.position_embeddings[
+                cache_index, :
+            ]
+        x = token_embedding + position_embedding
```
Comment (mattdangerw) [inline keras_nlp/models/gpt2/gpt2_causal_lm.py:244]: I would actually give a name to the `add` layer from the base class, and use that layer here. To keep the logic of the forward pass as close as possible.

File: keras_nlp/models/gpt2/gpt2_causal_lm.py [hunk 17]
```diff
@@ -200,17 +213,138 @@ def backbone_cls(cls):
...
+            x: a dense float Tensor, the next token logits of `inputs`.
+            cache: a dense float Tensor, the updated cache.
+        """
+        token_ids = inputs["token_ids"]
+        padding_mask = inputs["padding_mask"]
+        token_embedding = self.token_embeddings(token_ids)
+        if cache_index is None:
+            position_embedding = self.position_embeddings(token_embedding)
+        else:
+            position_embedding = self.position_embeddings.position_embeddings[
```
Comment (mattdangerw) [inline keras_nlp/models/gpt2/gpt2_causal_lm.py:241]: We should probably find a way to move this logic into the position embedding itself, rather than sometimes calling the layer, sometimes not.

File: keras_nlp/samplers/sampler.py [hunk 18]
```diff
@@ -39,6 +40,17 @@
...
+        `keras_nlp.layers.CachedMultiHeadAttention` layer to make the
+        decoding faster by avoiding duplicated computation. See more
+        details in the docstring of
+        `keras_nlp.layers.CachedMultiHeadAttention`.
+    token_probs: a dense float tensor of shape
```
Comment (mattdangerw) [inline keras_nlp/samplers/sampler.py:49]: This still feels pretty confusing as an arg.

As a user, I would think. We have a token_probability_fn and a token_probs pass in? What is the point of passing a function for compute the next tokens if I also have to pass them in.

I'm pretty sure we can avoid adding this arg entirely, will play around.

File: keras_nlp/layers/cached_multi_head_attention.py (hunk 2 above)
Comment (mattdangerw) [reply keras_nlp/layers/cached_multi_head_attention.py:66]: Also, won't this mean in our current forward pass we will compute a causal mask twice and merge the same thing twice. That feels both confusing and inefficient!

**Review by mattdangerw — CHANGES_REQUESTED** [review]
Thanks! This is shaping up!

There are a few unaddressed comments from an earlier round of review, added another pass now.

Still want to play around with our underlying graph a bit more, so may add a few more as I go.

File: keras_nlp/models/gpt2/gpt2_causal_lm.py [hunk 19]
```diff
@@ -200,17 +213,138 @@ def backbone_cls(cls):
...
+        x = {
+            "token_ids": token_ids,
+            "padding_mask": padding_mask,
+        }
+
+        batch_size = tf.shape(token_ids)[0]
+        outputs = tf.zeros(
+            [batch_size, max_length, self.backbone.vocabulary_size]
+        )
+        cache = tf.zeros(
```
Comment (mattdangerw) [inline keras_nlp/models/gpt2/gpt2_causal_lm.py:300]: Another thing I just remembered, we should update this cache shape so that `batch_size` comes first (will require updates elsewhere, e.g. stack/unstack axes).

The issue is that these axis choices are getting reflected all the way down to the cache layer, and even though we are not doing it today, we should support someone else writing a functional cache model. For functional models, the batch axis must come first.

File: keras_nlp/models/gpt2/gpt2_causal_lm.py [hunk 20]
```diff
@@ -235,10 +371,26 @@ def generate(
             # backward compat.
             sampler.jit_compile = self.jit_compile
         sampler.run_eagerly = self.run_eagerly
+        initial_output, cache = None, None
+        x, _, _ = self.preprocessor(prompt)
```
Comment (mattdangerw) [inline keras_nlp/models/gpt2/gpt2_causal_lm.py:375]: Another thing I just noticed, this is assuming that the preprocessor has a wider sequence length than max_length, I'm not sure we should make that assumption, as `preprocessor` is passed in by the user.

**Comment by mattdangerw** [comment]
@chenmoneygithub also a bug!

If you pass a variable length batch, e.g. `print(gpt2_lm.generate(["that's weird", "that's even weirder"], sampler="greedy", max_length=30))`, the longer sentence wont be respected. The word "weirder" will get deleted in the output.

File: keras_nlp/layers/cached_multi_head_attention.py (hunk 2 above)
Comment (fchollet) [reply keras_nlp/layers/cached_multi_head_attention.py:66]: Surely the call method's signature should be identical to that of the parent except for cache related args.

File: keras_nlp/models/gpt2/gpt2_backbone.py (hunk 5 above)
Comment (fchollet) [reply keras_nlp/models/gpt2/gpt2_backbone.py:147]: Black will go with the former version since the line is short enough

File: keras_nlp/models/gpt2/gpt2_causal_lm.py (hunk 14 above)
Comment (fchollet) [reply keras_nlp/models/gpt2/gpt2_causal_lm.py:194]: An alternative, if you are so inclined, is to move those setattr calls to a new method (`set_layer_attributes()`?) decorated with a do-not-track decorator. Though the `@property` accessors are fine too.

File: keras_nlp/models/gpt2/gpt2_causal_lm.py [hunk 21]
```diff
@@ -200,17 +213,138 @@ def backbone_cls(cls):
...
         return GPT2CausalLMPreprocessor
 
-    def _get_token_probability(self, prompt, mask):
+    def call_with_cache(self, inputs, cache, cache_index=None):
+        """Forward pass of `GPT2CausalLM` with cache.
+
+        The difference between `call_with_cache` and normal `__call__` is in
+        this method, a `cache` arg is set, and the inputs is of
+        `sequence_length=1`. By cachine the previous key/value in multi-head
+        attention, we avoid recomputing the outputs of seen tokens.
```
Comment (fchollet) [inline keras_nlp/models/gpt2/gpt2_causal_lm.py:222]: The docstrings might need a round of copyedits in general

**Review by fchollet — COMMENTED** [review]
This mechanism is fine -- easy to read and maintain 👍

File: keras_nlp/layers/cached_multi_head_attention.py (hunk 2 above)
Comment (mattdangerw) [reply keras_nlp/layers/cached_multi_head_attention.py:66]: We should keep the default behavior one to one with the parent except for caching behavior.

I care less about supporting all the bells and whistles on the base class (e.g. supporting implicit masks with the cache might take some thought, and we have no direct use case yet).

Comment (chenmoneygithub) [reply keras_nlp/layers/cached_multi_head_attention.py:66]: This is directly from the MHA: [link](https://github.com/keras-team/keras/blob/e6784e4302c7b8cd116b74a784f4b78d60e83c26/keras/layers/attention/multi_head_attention.py#L550)

Should we stick to the parent class for now and reevaluating if `use_causal_mask` is required for MHA, if it is then I guess we should keep it for consistency.

File: keras_nlp/models/gpt2/gpt2_backbone.py (hunk 6 above)
Comment (chenmoneygithub) [reply keras_nlp/models/gpt2/gpt2_backbone.py]: I was doing this way to avoid the overhead of massive `get_layer` calls, and for some readability. Also I am unsure how the layer is recognized, by doing this way we can still load the old checkpoints.

But anyway, it is fine to me to move these `get_layer` into `call_with_cache`, which should not significantly affect the performance.

File: keras_nlp/models/gpt2/gpt2_backbone.py (hunk 7 above)
Comment (chenmoneygithub) [reply keras_nlp/models/gpt2/gpt2_backbone.py]: nice idea.

File: keras_nlp/models/gpt2/gpt2_backbone.py (hunk 6 above)
Comment (mattdangerw) [reply keras_nlp/models/gpt2/gpt2_backbone.py]: > Also I am unsure how the layer is recognized, by doing this way we can still load the old checkpoints.

It is a bit complicated, we can talk offline. Basically you won't see any issues `backbone.save_weights`/`backbone.load_weights`, but your changes (as currently stand on the PR) would change `causal_lm.save_weights`/`causal_lm.load_weights` because the causal lm will think it has all these extra layers!

Another way to see this is just printing `causal_lm.summary()`. The causal lm will print a backbone with `124m` parameters, but then duplicate all the individual layers in the summary again.

Happy to talk through more, but the short version is we can either do `get_layer("name")` calls, or the setattr tricks Francois mentioned below. `get_layer("name")` call straight in `call_with_cache` seems the simplest to get this landed.

Comment (chenmoneygithub) [reply keras_nlp/models/gpt2/gpt2_backbone.py]: thx for the explanation, i will refactor the code.

File: keras_nlp/samplers/beam_sampler.py (hunk 9 above)
Comment (chenmoneygithub) [reply keras_nlp/samplers/beam_sampler.py:103]: let's do a followup, and we can still export beam without cache.

File: keras_nlp/layers/cached_multi_head_attention.py (hunk 12 above)
Comment (chenmoneygithub) [reply keras_nlp/layers/cached_multi_head_attention.py:88]: so... this is written in this way because the original prompt is padded, and the special case is not `cache_index=0`, but the first pass. 

I don't think I fully understand this comment tho, what do you mean by "slice by `cache_index + seq_len`"? a code snippet might help.

File: keras_nlp/layers/cached_multi_head_attention.py (hunk 13 above)
Comment (chenmoneygithub) [reply keras_nlp/layers/cached_multi_head_attention.py:90]: simply doing stack to unify the `update_indices ` for k and v isn't very helpful to me.

File: keras_nlp/models/gpt2/gpt2_causal_lm.py (hunk 16 above)
Comment (chenmoneygithub) [reply keras_nlp/models/gpt2/gpt2_causal_lm.py:244]: sounds good, refactored.

File: keras_nlp/models/gpt2/gpt2_causal_lm.py (hunk 17 above)
Comment (chenmoneygithub) [reply keras_nlp/models/gpt2/gpt2_causal_lm.py:241]: I can open an api for it as a followup, would like to split it to make the commit clean.

File: keras_nlp/samplers/sampler.py (hunk 18 above)
Comment (chenmoneygithub) [reply keras_nlp/samplers/sampler.py:49]: talked offline, let's remove it for now, if contrastive search needs it, let's add it back

File: keras_nlp/models/gpt2/gpt2_causal_lm.py (hunk 20 above)
Comment (chenmoneygithub) [reply keras_nlp/models/gpt2/gpt2_causal_lm.py:375]: I think it's okay when preprocessor has a shorter seq length. The cache's length is `max(max_length, gpt2_seq_len)`, if the preprocessor has a `seq_len < max_length < gpt2_seq_len`, then inside the `sampler` call, the prompt will be padded to `max_length`. If `seq_len > gpt2_seq_len`, the forward pass is broken anyway.

File: keras_nlp/layers/cached_multi_head_attention.py (hunk 13 above)
Comment (mattdangerw) [reply keras_nlp/layers/cached_multi_head_attention.py:90]: Let me share some code. May make sense to cover this in a follow up.

**Review by mattdangerw — COMMENTED** [review]
Thanks for the changes!

File: keras_nlp/models/gpt2/gpt2_causal_lm.py [hunk 22]
```diff
@@ -200,17 +201,142 @@ def backbone_cls(cls):
...
+
+        Args:
+            token_ids: a dense int Tensor, input token ids.
+            padding_mask: a dense bool Tensor, input padding mask.
+            cache: a dense float Tensor, the cache of key and value.
+            cache_index: int, or int Tensor, defaults to None. If set, it
+                represents the index of current inputs in the whole sequence.
+
+        Returns:
+            x: a dense float Tensor, the next token logits of `inputs`.
```
Comment (mattdangerw) [inline keras_nlp/models/gpt2/gpt2_causal_lm.py:220]: Same issue here as with TransformerDecoder. Rewrite this to reference the tuple structure, the name of the tuple outputs in our implementation should be totally irrelevant. E.g.

A `(logits, cache)` tuple. Where the first output is the language model logits for the input `token_ids` and the second output is the updated cache.

File: keras_nlp/models/gpt2/gpt2_causal_lm.py [hunk 23]
```diff
@@ -200,17 +201,142 @@ def backbone_cls(cls):
...
+        max_length, `self.num_heads`, `self.hidden_dim // self.num_heads`].
+        The first dim represents it's a key or value in multi-head attention.
+
+        Args:
+            initial_inputs: a dense Tensor, the initial inputs to the decoding
+                loop.
+            max_length: int, the max length of the generated sequence.
+
+        Returns:
+            cache: a dense float Tensor, the cache of key and value.
```
Comment (mattdangerw) [inline keras_nlp/models/gpt2/gpt2_causal_lm.py:277]: Here as well, update the doc.

**Review by mattdangerw — COMMENTED** [review]
Couple more comments.

File: keras_nlp/layers/cached_multi_head_attention.py (hunk 13 above)
Comment (mattdangerw) [reply keras_nlp/layers/cached_multi_head_attention.py:90]: Here's some code showing where we could head for a simplified implementation of `CachedMultiHeadAttention`. https://github.com/mattdangerw/keras-nlp/commit/a4a9e4ba3c1e5849b07defc91c55a848cec9d73e. Let's discuss offline, but probably we can just do this as a follow up.

Comment (mattdangerw) [reply keras_nlp/layers/cached_multi_head_attention.py:90]: Talked offline. Let's just do this as a follow up. Getting the mask right is a bit tricky, so we can leave that for later.

File: keras_nlp/layers/cached_multi_head_attention.py (hunk 12 above)
Comment (mattdangerw) [reply keras_nlp/layers/cached_multi_head_attention.py:88]: Let's do this as a follow up!

File: keras_nlp/layers/cached_multi_head_attention.py (hunk 11 above)
Comment (mattdangerw) [reply keras_nlp/layers/cached_multi_head_attention.py]: Will do this on a follow up

File: keras_nlp/models/gpt2/gpt2_causal_lm.py (hunk 17 above)
Comment (mattdangerw) [reply keras_nlp/models/gpt2/gpt2_causal_lm.py]: sounds good!

File: keras_nlp/layers/cached_multi_head_attention.py (hunk 2 above)
Comment (chenmoneygithub) [reply keras_nlp/layers/cached_multi_head_attention.py:66]: talked offline, we are flipping `use_causal_mask=False` for now, and put the rest into a followup PR.

File: keras_nlp/models/gpt2/gpt2_causal_lm.py (hunk 19 above)
Comment (chenmoneygithub) [reply keras_nlp/models/gpt2/gpt2_causal_lm.py:291]: sg, let me change it.

File: keras_nlp/models/gpt2/gpt2_causal_lm.py [hunk 24]
```diff
@@ -200,17 +201,142 @@ def backbone_cls(cls):
...
+            cache,
+            cache_index,
+        )
+        return output, cache
 
     def generate(
         self,
         prompt,
         max_length,
+        use_cache=True,
```
[the remark this answers is not in the document]
Comment (chenmoneygithub) [reply keras_nlp/models/gpt2/gpt2_causal_lm.py:339]: sounds good!

File: keras_nlp/layers/cached_multi_head_attention_test.py [hunk 25]
```diff
@@ -0,0 +1,116 @@
...
+
+    def test_cache_call_with_tf_while(self):
+        batch_size = 2
+        seq_len = 5
+        num_heads = 2
+        key_dim = 4
+
+        layer = CachedMultiHeadAttention(num_heads=num_heads, key_dim=key_dim)
+        x = tf.random.uniform(shape=[batch_size, seq_len, num_heads * key_dim])
+        # [2, batch_size, seq_length, num_heads, key_dim]
```
Comment (mattdangerw) [inline keras_nlp/layers/cached_multi_head_attention_test.py:75]: this comment is just lifted straight from the line below, delete it

**Review by mattdangerw — APPROVED** [review]
Thanks! Approving!

Mainly some copy edits. Though I do think we might want to figure out the one unresolved comment change re preprocessor.sequence_length messing up in some edge cases.

File: keras_nlp/layers/cached_multi_head_attention_test.py [hunk 26]
```diff
@@ -0,0 +1,116 @@
...
+)
+
+
+class CachedMultiHeadAttentionTest(tf.test.TestCase, parameterized.TestCase):
+    def test_valid_call(self):
+        layer = CachedMultiHeadAttention(num_heads=2, key_dim=4)
+        x = tf.random.uniform(shape=[2, 2, 8])
+        layer(query=x, value=x)
+
+    def test_cache_call_is_correct(self):
```
[the remark this answers is not in the document]
Comment (chenmoneygithub) [reply keras_nlp/layers/cached_multi_head_attention_test.py:31]: sg

**Comment by chenmoneygithub** [comment]
@mattdangerw Thanks!

Re the `seq_len` of preprocessor - to me it's users' liability to choose the right `seq_len` of `preprocessor`, otherwise they will also hit performance issue if they just want to use the backbone itself. For most of our users, they probably won't hit this issue. I am also fine with swapping the `sequence_length` to `gpt2.sequence_length` before calling `preprocessor`.

Actually as we talked offline, we might want to rethink our logic on calling `preprocessor` and `tokenizer` inside the `generate` method.

File: keras_nlp/models/gpt2/gpt2_causal_lm.py (hunk 21 above)
Comment (chenmoneygithub) [reply keras_nlp/models/gpt2/gpt2_causal_lm.py:210]: applied

**Comment by mattdangerw** [comment]
> Actually as we talked offline, we might want to rethink our logic on calling preprocessor and tokenizer inside the generate method.

sgtm to revisit this after looking at simplifying the preprocessing generally for `generate()`. Thanks!

**Review by fchollet — COMMENTED** [review]
Everything looks good to me! 👍

**Comment by chenmoneygithub** [comment]
Thanks all for reviewing this looooong PR! Cheers.
